src/match.ts

```diff
@@ -146,3 +147,227 @@ export function getRightmostFailure(match: MatchFail): MatchFail {
+    const scope = rightmostFailure.scope;
+    if (scope.stream) {
+      const path = scope.stream.path;
+      output.push(`Position: ${path.toString()}\n`);
+
+      // Try to show the value at this position
+      if (scope.stream.value !== undefined) {
+        output.push(`Current value: ${formatValue(scope.stream.value)}\n`);
+      } else if (!scope.stream.done) {
+        output.push(`Current value: <no value>\n`);
+      } else {
+        output.push(`Current value: <end of input>\n`);
+      }
+    }
+  }
+
+  output.push("\n=== End Visualization ===");
+  return output.join("");
+}
+
+function getPatternName(pattern: Pattern): string {
+  switch (pattern.kind) {
+    case PatternKind.Reference:
+      return pattern.name;
+    case PatternKind.Pipeline:
+      return "Pipeline";
+    case PatternKind.Then:
+      return "Then";
+    case PatternKind.Or:
+      return "Or";
+    case PatternKind.And:
+      return "And";
+    case PatternKind.Maybe:
+      return "Maybe";
+    case PatternKind.Slice:
+      return "Slice";
+    case PatternKind.Equal:
+      return `Equal(${JSON.stringify(pattern.value)})`;
+    case PatternKind.Includes:
+      return "Includes";
+    case PatternKind.Range:
+      return "Range";
+    case PatternKind.RegExp:
+      return `RegExp(${pattern.pattern})`;
+    case PatternKind.Type:
+      return `Type(${pattern.type})`;
+    case PatternKind.Variable:
+      return `Variable(${pattern.name})`;
+    case PatternKind.Run:
+      return pattern.name ? `Run(${pattern.name})` : "Run";
+    case PatternKind.Character:
+      return `Character(${pattern.characterClass})`;
+    case PatternKind.Special:
+      return `Special(${pattern.name})`;
+    case PatternKind.Into:
+      return "Into";
+    case PatternKind.Over:
+      return "Over";
+    case PatternKind.Not:
+      return "Not";
+    case PatternKind.Any:
+      return "Any";
+    case PatternKind.Ok:
+      return "Ok";
+    case PatternKind.Fail:
+      return "Fail";
+    case PatternKind.End:
+      return "End";
+    default:
+      return "Unknown";
+  }
+}
+
+function isPipeline(pattern: Pattern): boolean {
+  return pattern.kind === PatternKind.Pipeline;
+}
+
+function formatValue(value: unknown): string {
+  if (value === undefined) return "<undefined>";
+  if (value === null) return "<null>";
+
+  if (typeof value === "string") {
+    return `"${value.length > 50 ? value.substring(0, 50) + "..." : value}"`;
+  }
+
+  if (typeof value === "number" || typeof value === "boolean") {
+    return String(value);
+  }
+
+  if (Array.isArray(value)) {
+    if (value.length === 0) return "[]";
+    if (value.length > 3) {
+      return `[${
+        value.slice(0, 3).map(formatValue).join(", ")
+      }, ... (${value.length} items)]`;
+    }
+    return `[${value.map(formatValue).join(", ")}]`;
+  }
+
+  if (typeof value === "object") {
+    const obj = value as Record<string, unknown>;
+    if ("kind" in obj) {
```

Comment:
Lets not add this special case, just format all objects in a consistent way. Some patterns are themselves too large to just stringify fully.